packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts

```diff
@@ -664,8 +664,26 @@ export class OdspDocumentDeltaConnection extends DocumentDeltaConnection {
 				super.addTrackedListener(
 					event,
 					(msg: ISignalMessage | ISignalMessage[], documentId?: string) => {
-						if (!this.enableMultiplexing || !documentId || documentId === this.documentId) {
+						const msgs = Array.isArray(msg) ? msg : [msg];
+						if (!this.enableMultiplexing || !documentId) {
+							if (!documentId) {
+								assert(
+									msgs.every((m) => !m.targetClientId),
+									"targetClientId defined while documentId is not",
+								);
+							}
 							listener(msg, documentId);
+							return;
+						}
+
+						const filteredMsgs = msgs.filter(
+							(m) =>
+								documentId === this.documentId &&
```

Comment:
Let's pull this out. Have:
```typescript
if (documentId !== this.documentId) {
   return;
}
```
It will be a lot more common that that rejects or allows every message. So faster to use one off check before per message filtering.

Otherwise, looks correct.